src/harness/reference_models/iap/iap.py

```diff
@@ -0,0 +1,623 @@
+  is applicable for FSS Co-Channel and ESC Sensor protection points.Also for 
+  protection points within PPA and GWPZ protection areas. Post IAP grant EIRP
+  and interference contribution is calculated over each 5MHz channel
+
+  Args:
+    protection_point: A protection point as a tuple (longitude, latitude).
+    channels: List of tuples containing (low_freq, high_freq) of each 5MHz 
+              segment of the whole frequency range of protected entity
+    low_freq: low frequency of protection constraint (Hz)
+    high_freq: high frequency of protection constraint (Hz)
+    grant_objects: a list of grant requests, each one being a tuple 
+                   containing grant information
+    fss_info: contains fss point and antenna information
+    esc_antenna_info: contains information on ESC antenna height, azimuth, 
+                      gain and pattern gain
+    threshold: protection threshold (mW)
+    protection_ent_type: an enum member of class ProtectedEntityType 
+    asas_interference: object of class type AggregateInterferenceOutputFormat 
+    aggregate_interference: object of class type AggregateInterferenceOutputFormat 
+  Returns:
+    Updates asas_interference instance of class type AggregateInterferenceOutputFormat.
+    It is updated in the format of {latitude : {longitude : [ asas_p(mW/IAPBW),.. ]}}
+    and aggregate_interference instance of class type AggregateInterferenceOutputFormat.
+    It is updated in the format of {latitude : { longitude : [a_p(mW/IAPBW),..] }}
+  """
+  
+  # Get all the grants inside neighborhood of the protection entity
+  grants_inside = interf.findGrantsInsideNeighborhood(
+                    grant_objects, protection_point, protection_ent_type)
+   
+  if len(grants_inside) > 0:
+    # Get protection constraint of protected entity
+    protection_constraint = interf.ProtectionConstraint(latitude=protection_point[1],
+                               longitude=protection_point[0], low_frequency=low_freq,
+                               high_frequency=high_freq, entity_type=protection_ent_type)
+
+    # Get all the grants fully/partially overlap with protection point,p
+    neighborhood_grants = interf.findOverlappingGrants(
+                            grants_inside, protection_constraint)
+
+    num_unsatisfied_grants = len(neighborhood_grants)
+    
+    # Get number of grants, IAP protection threshold and fairshare per channel 
+    if num_unsatisfied_grants > 0:
+      # algorithm to calculate interference within IAPBW using EIRP obtained for all 
+      # the grants through application of IAP
+
+      # Initialize list of aggregate interference from all the grants 
+      # ( including grants from managing and peer SAS )
+      aggr_interf = [0] * len(channels)
+      # Initialize list of aggregate interference from all the grants 
+      # ( including only grants from managing SAS )
+      asas_interf = [0] * len(channels)
```

Comment:
As mentioned before, I think you need to move aggr_interf & asas_interf to the very beginning of this method.

The idea is that you need to calculate them even if there is no associated grants for this protected point.

Reply:
Moved   outside